opentelemetry-api/src/opentelemetry/trace/sampling.py

```diff
@@ -0,0 +1,128 @@
+# Copyright 2019, OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import abc
+from typing import Dict, Mapping, Optional, Sequence
+
+# pylint: disable=unused-import
+from opentelemetry.trace import Link, SpanContext
+from opentelemetry.util.types import AttributeValue
+
+
+class Decision:
+    """A sampling decision as applied to a newly-created Span.
+
+    Args:
+        sampled: Whether the `Span` should be sampled.
+        attributes: Attributes to add to the `Span`.
+    """
+
+    def __repr__(self) -> str:
+        return "{}({}, attributes={})".format(
+            type(self).__name__, str(self.sampled), str(self.attributes)
+        )
+
+    def __init__(
+        self,
+        sampled: bool = False,
+        attributes: Mapping[str, "AttributeValue"] = None,
+    ) -> None:
+        self.sampled: bool
+        self.attributes: Dict[str, "AttributeValue"]
+
+        self.sampled = sampled
+        if attributes is None:
+            self.attributes = {}
+        else:
+            self.attributes = dict(attributes)
+
+
+class Sampler(abc.ABC):
+    @abc.abstractmethod
+    def should_sample(
+        self,
+        parent_context: Optional["SpanContext"],
+        trace_id: int,
+        span_id: int,
+        name: str,
+        links: Optional[Sequence["Link"]] = None,
+    ) -> "Decision":
+        pass
+
+
+class StaticSampler(Sampler):
+    """Sampler that always returns the same decision."""
+
+    def __init__(self, decision: "Decision"):
+        self.decision = decision
+
+    def should_sample(
+        self,
+        parent_context: Optional["SpanContext"],
+        trace_id: int,
+        span_id: int,
+        name: str,
+        links: Optional[Sequence["Link"]] = None,
+    ) -> "Decision":
+        return self.decision
+
+
+class ProbabilitySampler(Sampler):
+    def __init__(self, rate: float):
+        self._rate = rate
+        self._bound = self.get_bound_for_rate(self._rate)
+
+    # The sampler checks the last 8 bytes of the trace ID to decide whether to
+    # sample a given trace.
+    CHECK_BYTES = 0xFFFFFFFFFFFFFFFF
+
+    @classmethod
+    def get_bound_for_rate(cls, rate: float) -> int:
+        return round(rate * (cls.CHECK_BYTES + 1))
```

Comment:
If rate is less than `0.0000000000000000000271`, bound will be rounded to 0 which means it never gets sampled.
Do we consider to round up or this is just too pedantic?

Reply:
Actually it is `0.0000000000000000000813` since bound is rounded to `1` (and all-zero trace id is illegal according to the TraceContext spec). 😄

Reply:
I used `round` intentionally here for what I thought were pedantic reasons: this way each of the `2^64` sampling rate "buckets" are split exactly in half:

```
            r <= 2^-65     : 0
    2^-65 < r <= 3 * 2^-65 : 1
3 * 2^-65 < r <= 5 * 2^-65 : 2
```

and etc.

 I think if you pass in a sampling rate `<= 2^-65` it's more correct for us to make the actual rate `0` than `2^-64`.

This doesn't exactly work because float precision is worse than `2^-64` for values close to `1`, but it works near `0` and is at least theoretically correct (or should be, there may still be bugs here).

> all-zero trace id is illegal according to the TraceContext spec.

Oh that's interesting, so the trace ID space is `[0x1, 0xf...f]`, not `[0x0, 0xf...f]`. So would you expect this line to be:

```python
return 1 + round(rate * cls.CHECK_BYTES)
```

I think that'd give us the same behavior as above, but always sample trace ID `0x0`. 

Should we have special handling for `0x0` trace IDs?

Reply:
I'm also happy to leave this as-is and treat `0x0` as a regular ID here, but treat it differently at span creation time. Up to you. :D

Reply:
I don't have any preference here as I cannot think of a scenario where people would need such low but non-zero sample rate (unless someone is instrumenting [SETI@home](https://setiathome.berkeley.edu)). 😆

Reply:
Something I missed above: there are and `2^64 - 1` trace IDs besides `0x0` that end in `0x0000000000000000`. If we want to always/never sample the all-zero trace ID we have to make that decision before masking.